### PR TITLE
docs(dev) plugin development guide

### PR DESCRIPTION
**Pull request by thibaultCha**
Add a guide for plugin development, covering all aspects (such as event
handlers, DAO access, Admin API extension, etc...)

Also introduce several other changes to provide better docs and building
tools:

- The "Public Lua API Reference": generated from Kong's source code
  documentation and ldoc, a custom template and some hacks allow for a
  documentation of the available Kong Lua modules. Unfortunately for
  now, not all modules are documented, nor properly documented, but the
  most important ones (linked from the plugin dev guide) are. `@see` links
  are also broken for now.
- This Reference is built from a `gulp` task and requires some manual
  work.
- A tiny Jekyll plugin to create a "book". A book is a linked list of
  chapters. Each chapter is a markdown page. next, previous, and other
  chapters links are inserted into a page's metadata a build time.
- Documentation for the `kong migrations` command.

Implement #107 and depends on Mashape/kong#657

---

I really can't waste more time on this, we will iterate over time. As mentioned, some things are not perfect yet and would require more work (and hacks, especially the Lua Reference) but that'll be enough for a first version I think.

@ahmadnassri @Nijikokun Could you please review the actual guide please? In terms of content/grammar. The files are in `app/docs/0.5.x/plugin-development`.

### CODE REVIEW

File: app/_layouts/docs.html [hunk 1]
```diff
@@ -52,13 +52,15 @@
...
         <li>
-          <a href="{{docs_url | append: item.url}}" {% if page.url contains item.url %}class="active"{% endif %}>{{item.text}}</a>
+          <a href="{{docs_url | append: item.url}}" {% if page.url == url %}class="active"{% endif %}>{{item.text}}</a>
           {% if item.items %}
           <ul>
             {% for subitem in item.items %}
+            {% capture sub_url %}{{ docs_url | append: subitem.url }}/{% endcapture %}
             <li>
-              <a href="{{docs_url | append: subitem.url}}">{{subitem.text}}</a>
+              <a href="{{docs_url | append: subitem.url}}" {% if page.url == sub_url %}class="active"{% endif %}>{{subitem.text}}</a>
```
Comment (ahmadnassri) [inline app/_layouts/docs.html]: `===`

File: app/_layouts/docs.html [hunk 2]
```diff
@@ -52,13 +52,15 @@
       {% endif %}
       <ul>
         {% for item in nav.items %}
+        {% capture url %}{{ docs_url | append: item.url }}/{% endcapture %}
         <li>
-          <a href="{{docs_url | append: item.url}}" {% if page.url contains item.url %}class="active"{% endif %}>{{item.text}}</a>
+          <a href="{{docs_url | append: item.url}}" {% if page.url == url %}class="active"{% endif %}>{{item.text}}</a>
```
Comment (ahmadnassri) [inline app/_layouts/docs.html]: `===`

**Comment by thibaultCha** [comment]
Also going to need help on how to organize this now:

screenshot 2015-10-28 15 44 26

Should the "Write logic" tile be higher? Should we have another icon for it (@jitachi @sinzone)? And find an ok text, I'll keep thinking about this.

**Comment by ahmadnassri** [comment]
installation, quick start, admin API, plugins, config, cli, proxy

**Comment by jitachi** [comment]
I'd suggest reorganize the area by highlighting what's more used. Like 2 banners for example for Installation and Quick Start.

**Comment by thibaultCha** [comment]
Updated. Also I went for:

> Looking for something Kong does not do for you? Easy: write it as a plugin. Learn how to write your own plugins for Kong.

on the plugin development tile.

**Comment by jitachi** [comment]
/docs needs to be redesigned, that first page, not even the sidebar navigation can handle it: 
image

**Comment by thibaultCha** [comment]
> /docs needs to be redesigned, that first page, not even the sidebar navigation can handle it:

+1

**Comment by ahmadnassri** [comment]
the side bar should fold for sections not currently in view.

**Comment by sinzone** [comment]
we need to start dogfooding and move it under gelato. @nikz

**Comment by thibaultCha** [comment]
This is not related to the admin API, and too far from it to even be in a gelato guide.

What is related to the admin API, yes.

**Comment by thibaultCha** [comment]
Thanks! Updated and now deploying this.

**Pull request merged**